auth-gateway/index.js

```diff
@@ -0,0 +1,162 @@
+// since signed, but that the issuer was Auth0 (since they hold the private key)
+const jwksFn = jwksRsa({
+  cache: true,
+  rateLimit: true,
+  jwksRequestsPerMinute: 5,
+  jwksUri: `https://${AUTH0_DOMAIN}/.well-known/jwks.json`
+});
+
+// Ref: https://github.com/auth0/node-jwks-rsa/blob/master/src/integrations/express.js
+function getSecretKey(header, cb) {
+  jwksFn.getSigningKey(header.kid, (err, key) => {
+    if (err) {
+      cb(err, null);
+      return;
+    }
+
+    cb(null, key.publicKey || key.rsaPublicKey);
+  });
+}
+
+// Verifies signature, as well as the issuer and audience, to guarantee
+// that the token was truly for our resource server
+const verifyToken = token =>
+  new Promise((resolve, reject) => {
+    jwt.verify(
+      token,
+      getSecretKey,
+      {
+        issuer: AUTH0_ISSUER,
+        audience: AUTH0_AUDIENCE,
+        clockTolerance: 2, // seconds to deal with clock skew
+        algorithms: ['RS256']
+      },
+      (err, dToken) => {
+        if (err) {
+          console.error(err);
+          return reject(err);
+        }
+
+        return resolve(dToken);
+      }
+    );
+  });
+
+const bearerPrefix = 'Bearer ';
+const bearerPrefixLen = bearerPrefix.length;
+
+const cookiePrefix = 'access_token=';
+const cookieOffset = cookiePrefix.length;
+
+const getAuthToken = req => {
+  let whereFoundCount = 0;
+
+  let bearerIdx;
+  let cookieIdx;
+
+  if (req.headers.cookie) {
+    cookieIdx = req.headers.cookie.indexOf(cookiePrefix);
+
+    if (cookieIdx > -1) {
+      whereFoundCount += 1;
+    }
+  }
+
+  if (req.query.access_token) {
+    whereFoundCount += 1;
+  }
+
+  if (req.headers.authorization) {
+    bearerIdx = req.headers.authorization.indexOf(bearerPrefix);
+
+    if (bearerIdx > -1) {
+      whereFoundCount += 1;
+    }
+  }
+
+  // oauth2 spec specifies token should be presented in only one location
+  if (whereFoundCount === 0) {
+    return null;
+  }
+
+  if (whereFoundCount > 1) {
+    console.warn('User specified > 1 access tokens', req.headers.origin);
+    return null;
+  }
+
+  // The method we are least likely to use; disable for now
+  // as this means access tokens are even easier to abuse
+  // if (req.query.access_token) {
+  //   return req.query.access_token;
+  // }
+
+  let token;
+
+  if (bearerIdx > -1) {
+    token = req.headers.authorization.substr(bearerPrefixLen);
+  } else if (cookieIdx > -1) {
+    token = req.headers.cookie.substr(cookieIdx + cookieOffset);
+  }
+
+  if (!token) {
+    return null;
+  }
+
+  const spaceIdx = token.indexOf(' ');
+
+  if (spaceIdx === -1) {
+    return token;
+  }
+
+  // When multiple authorization headers, they are delimited by a space and a ,
+  // https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers
+  // Similarly, when multiple cookies present, separated by value1, cookie2=value2
+  // or likewise by semicolon
+  return token.substr(0, spaceIdx - 1);
```

Comment:
This seems odd? The HTTP/1.1 [RFC 7235 4.2](https://tools.ietf.org/html/rfc7235#section-4.2) spec makes no mention of repeated authorization tokens (the "credentials" field is defined in [2.1] (https://tools.ietf.org/html/rfc7235#section-2.1)). Why would we want to support multiple authorization tokens at once, particularly if we're going to arbitrarily take the first one?

It seems better to treat everything after the `Bearer ` as the token, if the client screws up and specifies multiple authorization headers, then they'll get an authentication failure (because it certainly won't pass the signature check). This means clients won't rely on the first header being used (which seems like non-robust behavior, if the headers are accidentally interchanged later the code will start failing for non-obvious reasons).

Reply:
And if we are going to do this, it would seem we should do it only on the header token, not the cookie token (I believe `token`, here, could be either one).

Reply:
Multiple authorization tokens? I'm just specifying that there can be multiple cookies sent with the request. 

If you're referring to the possibility that the user sends the access_token in two places, I don't want this and check for it.

Reply:
From the spec you referenced: https://tools.ietf.org/html/rfc7230#section-3.2.2

https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers

```
A sender MUST NOT generate multiple header fields with the same field name in a message unless either the entire field value for that header field is defined as a comma-separated list [i.e., #(values)] or the header field is a well-known exception (as noted below).

A recipient MAY combine multiple header fields with the same field name into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field value to the combined field value in order, separated by a comma. The order in which header fields with the same field name are received is therefore significant to the interpretation of the combined field value; a proxy MUST NOT change the order of these field values when forwarding a message.
```

Reply:
To be more specific: I am explicitly not supporting multiple authorization tokens.

Reply:
Nope, you're correct. Authorization is not one of the headers that supports list syntax. I'll remove support there.

Reply:
Ok!